[AArch64] Improve urem by constant costs

davemgreen opened the pull request:
A urem by a constant, much like a udiv by a constant, can be expanded into a series of mul/add/shift instructions. The exact sequence of instructions depends on the constants and the types.

If the constant is a power-2 then a shift / and will be used, so the cost will be 1. This canonicalization happens relatively early so this likely has very little effect in practice (it does help the cost of funnel shifts).

For a non-power 2 the code for div will expand to a series of UMULH + Add + Shift + Add, depending on the constant. urem is generally udiv + mul + sub, so involves a few extra instructions. The UMULH is not always available, i32 will use umull+shift, and vector types will use umull+shift or umull+umull2+uzp depending on the vector size. v2i64 will be scalarized because there is no mul available. SVE does have a UMULH instruction.

The end result is that the costs should be closer to reality, with scalable types a little lower cost than the fixed-width versions. (In the future we might be able to use umulh for fixed-width when the SVE instruction is available, but for the moment this should favour scalable vectorization a little).

I've tried to make this patch only apply to constant UREM/UDIV instructions. SDIV and SREM are left until a later patch to prevent this becoming too complex. The funnel shift costs are changing as it believes it will need a urem to clamp the shift amount, which should be a power-2 value for most common types.

david-arm commented on llvm/lib/Analysis/TargetTransformInfo.cpp:
```diff
@@ -893,7 +893,8 @@ TargetTransformInfo::getOperandInfo(const Value *V) {
   // and check if the constant(s) are all powers of two.
-  if (isa<ConstantVector>(V) || isa<ConstantDataVector>(V)) {
+  if (isa<ConstantVector>(V) || isa<ConstantDataVector>(V) ||
+      isa<ConstantExpr>(V)) {
```
I wonder if it's worth committing this separately, or is it only exposed with your code changes? I know with the work @paulwalker-arm has been doing on migrating LLVM to use `splat(i32 ...)` for constant splats we should be seeing `ConstantExpr` used a lot more in future.

davemgreen replied:
Yeah this was a relatively late addition whilst writing SVE tests. #122469 has it separated out.

sushgokh commented:
@davemgreen 
I was just looking at the examples with 3 different scenarios for the constant in my mind:
1. If the constant is pow-2
2. If the constant is uniform
3. If the constant is negative

So, essentially, its like having 8 possibilities for the constant with these 3 variables.

Now, things change pretty quickly if the constant is negative(I think I said that things dont change much if the constant is  negative on the other patch but I was wrong).
e.g.
```
for <4 x i32> :
a) positive constant: ushr
b) negative constant: ushr + umull2 + umull + uzp2 + ushr     (for certain cases)
```
The non-uniform case also brings in more complexity but not that much.

So, maybe, can be break this patch so that we conquer just few possibilities(from 8 possibilities) in every patch at a time ? This  will ensure precision in cost estimation and easy verification I believe.

davemgreen commented:
> @davemgreen I was just looking at the examples with 3 different scenarios for the constant in my mind:
> 
> 1. If the constant is pow-2
> 2. If the constant is uniform
> 3. If the constant is negative
> 
> So, essentially, its like having 8 possibilities for the constant with these 3 variables.
> 
> Now, things change pretty quickly if the constant is negative(I think I said that things dont change much if the constant is negative on the other patch but I was wrong). e.g.
> 
> ```
> for <4 x i32> :
> a) positive constant: ushr
> b) negative constant: ushr + umull2 + umull + uzp2 + ushr     (for certain cases)
> ```
> 
> The non-uniform case also brings in more complexity but not that much.
> 
> So, maybe, can be break this patch so that we conquer just few possibilities(from 8 possibilities) in every patch at a time ? This will ensure precision in cost estimation and easy verification I believe.

I believe for udiv/urem there are only really 2 cases - positive power-2 and everything else. The power-2 cases almost never come up as they are canonicalised early and are mostly added for completeness. Like you say I don't think there is a lot of difference between uniform and non-uniforms, it just takes the worst-case.

The exact cost will depend on the constant though, and we are currently keeping things simple and assuming no simplification happens. Some positive constants will simplify, some negative ones will, it is hard to tell without doing a lot of maths to see what the exact instructions will be. We could try to use DivisionByConstantInfo to get a more precise expansion and figure out what will simplify, but that feels a bit more complex than is helpful considering the other costs that could be improved. Maybe one for the future, if you don't want to look into it yourself.

sushgokh commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -3545,20 +3545,58 @@ InstructionCost AArch64TTIImpl::getArithmeticInstrCost(
+            getArithmeticInstrCost(Instruction::AShr, Ty, CostKind,
+                                   Op1Info.getNoProps(), Op2Info.getNoProps());
         return MulCost * 2 + AddCost * 2 + ShrCost * 2 + 1;
       }
```
Maybe this part can be removed. Division by non-constant, uniform divisor is giving scalar code. Check [here](https://godbolt.org/z/TMGMEn8x3)

davemgreen replied:
Hi - This is inside the top-level `if (Op2Info.isConstant()) {`, so I don't think that should be an issue. We are only trying to update constant costs in this patch, to keep it simpler. I've added some uniform tests in f08824b935434b91f7352904a25f6309f2b3e6bd to check.

I think this bit of code can probably be removed when sdiv gets added. For now I will re-add the isScalableVector check to make sure the sdiv scores don't change yet.

sushgokh replied:
ok. Thanks. Will remove this if(){...} when I update the sdiv/srem patch.

davemgreen replied:
Thanks

davemgreen committed:
[AArch64] Improve urem by constant costs

A urem by a constant, much like a udiv by a constant, can be expanded into a
series of mul/add/shift instructions. The exact sequence of instructions
depends on the constants and the types.

If the constant is a power-2 then a shift / and will be used, so the cost will
be 1. This canonicalization happens relatively early so this likely has very
little effect in practice (it does help the cost of funnel shifts).

For a non-power 2 the code for div will expand to a series of UMULH + Add +
Shift + Add, depending on the constant. urem is generally udiv + mul + sub, so
involves a few extra instructions. The UMULH is not always available, i32 will
use umull+shift, and vector types will use umull+shift or umull+umull2+uzp
depending on the vector size. v2i64 will be scalarized because there is no mul
available. SVE does have a UMULH instruction.

The end result is that the costs should be closer to reality, with scalable
types a little lower cost than the fixed-width versions. (In the future we
might be able to use umulh for fixed-width when the SVE instruction is
available, but for the moment this should favour scalable vectorization a
little).

I've tried to make this patch only apply to constant UREM/UDIV instructions.
SDIV and SREM are left until a later patch to prevent this becoming too
complex. The funnel shift costs are changing as it believes it will need a urem
to clamp the shift amount, which should be a power-2 value for most common
types.

sushgokh reviewed (approved):
LGTM
